MsCorePkg/MacAddressEmulationDxe/Test/FindMatchingSnpTest.c

```diff
@@ -0,0 +1,217 @@
+  )
+{
+  // Arrange
+  MAC_EMULATION_SNP_NOTIFY_CONTEXT MacContext;
+  EFI_HANDLE *DummyHandle;
+  EFI_SIMPLE_NETWORK_PROTOCOL ExpectedMatch;
+  EFI_SIMPLE_NETWORK_PROTOCOL *ActualMatch;
+
+  EFI_SIMPLE_NETWORK_MODE Mode;
+
+  Mode.State = EfiSimpleNetworkStopped;
+  ExpectedMatch.Mode = &Mode;
+
+  size_t sizeHandle = sizeof(*DummyHandle);
+
+  DummyHandle = AllocateZeroPool(sizeHandle);
+  will_return(LocateHandleBuffer, 1);
+  will_return(LocateHandleBuffer, DummyHandle);
+  will_return(LocateHandleBuffer, EFI_SUCCESS);
+
+  will_return(HandleProtocol, &ExpectedMatch);
+  will_return(HandleProtocol, EFI_SUCCESS);
+
+  // Act
+  ActualMatch = FindMatchingSnp(SnpSupportsMacEmuCheck, &MacContext);
+
+  // Assert
+  assert_null(ActualMatch);
+
+  return UNIT_TEST_PASSED;
+}
+
+/**
+  Unit test for FindMatchingSnp_ReturnsPreviouslyAssignedSnp_WhenMultipleSnpSupportMacEmu ()
+
+  @param[in]  Context    [Optional] An optional parameter that enables:
+                         1) test-case reuse with varied parameters and
+                         2) test-case re-entry for Target tests that need a
+                         reboot.  This parameter is a VOID* and it is the
+                         responsibility of the test author to ensure that the
+                         contents are well understood by all test cases that may
+                         consume it.
+
+  @retval  UNIT_TEST_PASSED             The Unit test has completed and the test
+                                        case was successful.
+  @retval  UNIT_TEST_ERROR_TEST_FAILED  A test case assertion has failed.
+**/
+UNIT_TEST_STATUS
+EFIAPI
+FindMatchingSnp_ReturnsPreviouslyAssignedSnp_WhenMultipleSnpSupportMacEmu(
+  IN UNIT_TEST_CONTEXT  Context
+  )
+{
+  // Arrange
+  MAC_EMULATION_SNP_NOTIFY_CONTEXT MacContext;
+  EFI_HANDLE *DummyHandle;
+  EFI_SIMPLE_NETWORK_PROTOCOL ExpectedMatch;
+  EFI_SIMPLE_NETWORK_PROTOCOL *ActualMatch;
+
+  EFI_SIMPLE_NETWORK_MODE Mode;
+
+  MacContext.Assigned = TRUE;
+  SetMem(&MacContext.PermanentAddress, NET_ETHER_ADDR_LEN, 0xAA);
+
+  Mode.State = EfiSimpleNetworkInitialized;
+  Mode.IfType = NET_IFTYPE_ETHERNET;
+  Mode.MacAddressChangeable = TRUE;
+  SetMem(&Mode.PermanentAddress, NET_ETHER_ADDR_LEN, 0xAA);
+  ExpectedMatch.Mode = &Mode;
+
+  will_return(SnpSupportsMacEmulation, TRUE);
+
+  DummyHandle = AllocateZeroPool(sizeof(*DummyHandle));
+  SetMem(DummyHandle, sizeof(*DummyHandle), 0x11);
+  will_return(LocateHandleBuffer, 2);
+  will_return(LocateHandleBuffer, DummyHandle);
+  will_return(LocateHandleBuffer, EFI_SUCCESS);
+
+  will_return(HandleProtocol, &ExpectedMatch);
+  will_return(HandleProtocol, EFI_SUCCESS);
+
+  // Act
+  ActualMatch = FindMatchingSnp(SnpSupportsMacEmuCheck, &MacContext);
+
+  // Assert
+  assert_ptr_equal(&ExpectedMatch, ActualMatch);
+
+  return UNIT_TEST_PASSED;
+}
+
+UNIT_TEST_STATUS
+EFIAPI 
+FindMatchingSnpTestSetup (
+  IN UNIT_TEST_CONTEXT  Context
+  )
+{
+  gBS = &mBootServices;
```

Comment:
Only the following 2 function pointers are initialized in the mBootServices structure.  If code being tested in the driver is changed to use other pointers, I assume we would try to de-reference a NULL.  Should we copy the mBootServices structure to our local copy first before re-assigning gBS and overriding the 2 functions?

Reply:
I'm not following you here.

In the unit test we would want the program to crash if we dereference a null pointer since that means the test isn't fully complete yet.

Not sure what point would be to copy mBootService to a local copy here since it's just a statically assigned memory range.